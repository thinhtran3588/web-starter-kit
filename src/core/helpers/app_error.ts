export class AppError extends Error {
  code: string;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    // Set the prototype explicitly.
    Object.setPrototypeOf(this, AppError.prototype);
  }
}
