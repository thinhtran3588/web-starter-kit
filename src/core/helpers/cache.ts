// eslint-disable-next-line @typescript-eslint/no-explicit-any
const cacheStorage: Record<string, any> = {};

const get = <T>(key: string): T => {
  return cacheStorage[key];
};

const clear = (key: string): void => {
  delete cacheStorage[key];
};

const expire = (key: string, time: number): void => {
  setTimeout(() => {
    clear(key);
  }, time);
};

const set = <T>(key: string, value: T, expireTime?: number): void => {
  cacheStorage[key] = value;
  if (expireTime !== undefined) {
    expire(key, expireTime);
  }
};

export const cache = {
  get,
  set,
  clear,
  expire,
};
