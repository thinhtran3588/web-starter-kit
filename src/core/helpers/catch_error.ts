/* eslint-disable @typescript-eslint/no-explicit-any */
import { handleError } from './handle_error';

export const catchError = <T>(
  action: T,
  setIsBusy?: (f: (draft: boolean) => boolean | void) => void,
  matchedCodes?: { [code: string]: string },
  ignoreCodes?: { [code: string]: true },
): T => {
  return ((async (...params: any[]): Promise<any> => {
    try {
      setIsBusy && setIsBusy(() => true);
      return await (action as any)(...(params as any));
    } catch (error) {
      handleError(error, matchedCodes, ignoreCodes);
    } finally {
      setIsBusy && setIsBusy(() => false);
    }
    return undefined;
  }) as unknown) as T;
};
