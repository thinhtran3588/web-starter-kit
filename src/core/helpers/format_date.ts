import dayjs from 'dayjs';
import { config } from '@app/config';
import { FieldValueType } from '../interfaces';

export const formatDate = (value: FieldValueType): string => {
  if (!value) {
    return '';
  }
  return dayjs(value).format(config.dateFormat);
};
