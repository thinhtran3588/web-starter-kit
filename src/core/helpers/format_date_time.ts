import dayjs from 'dayjs';
import { config } from '@app/config';
import { FieldValueType } from '../interfaces';

export const formatDateTime = (value: FieldValueType): string => {
  if (!value) {
    return '';
  }
  return dayjs.unix(parseFloat((value as string).substring(0, 10))).format(config.dateTimeFormat);
};
