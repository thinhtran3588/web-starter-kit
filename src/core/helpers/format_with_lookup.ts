import { FieldValueType, PickerDataItem } from '../interfaces';

export const formatWithLookup = (lookup: PickerDataItem<string>[]) => (value: FieldValueType) => {
  if (!value) {
    return '';
  }
  const loginType = lookup.find((m) => m.value === value);
  return loginType ? loginType.label : '';
};
