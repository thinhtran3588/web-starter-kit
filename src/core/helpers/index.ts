export * from './format_with_lookup';
export * from './format_date';
export * from './format_date_time';
export * from './show_notification';
export * from './catch_error';
export * from './handle_error';
export * from './app_error';
export * from './sanitize_form_data';
export * from './get_error_message';
export * from './sleep';
