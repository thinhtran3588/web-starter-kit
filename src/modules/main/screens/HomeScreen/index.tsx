import React from 'react';
import { Layout } from '@app/components';
import { WithTranslation, withTranslation } from '@app/core';
import { CoverImage, Features, Quote, Help } from './components';

type Props = WithTranslation;

const Screen = (props: Props): JSX.Element => {
  const { t } = props;
  return (
    <Layout title={t('home')} description={t('description')}>
      <CoverImage t={t} />
      <Features t={t} />
      <Quote t={t} />
      <Help t={t} />
    </Layout>
  );
};

Screen.getInitialProps = async () => {
  return {
    namespacesRequired: ['common', 'home'],
  };
};

export const HomeScreen = withTranslation('home')(Screen);
