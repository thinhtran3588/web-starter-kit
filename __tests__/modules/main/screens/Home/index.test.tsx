import React from 'react';
import { render } from '@testing-library/react';
import { HomeScreen } from '@app/modules/main/screens/HomeScreen';
import { TestBaseComponent } from '@test/helpers/TestBaseComponent';

describe('@app/modules/main/screens/HomeScreen', () => {
  beforeEach(() => {});

  it('renders successfully', async () => {
    const { baseElement } = render(
      <TestBaseComponent>
        <HomeScreen />
      </TestBaseComponent>,
    );
    expect(baseElement).toMatchSnapshot();
  });
});
