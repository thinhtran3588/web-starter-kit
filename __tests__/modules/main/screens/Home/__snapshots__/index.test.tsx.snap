// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@app/modules/main/screens/Home renders successfully 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionFixed makeStyles-root-3 MuiAppBar-colorPrimary mui-fixed"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <next-link
          href="/"
        >
          <a
            href="/"
            title="Home"
          >
            <img
              alt="Logo"
              class="makeStyles-logo-5"
              src="/static/images/logo.svg"
            />
          </a>
        </next-link>
        <div
          class="makeStyles-flexGrow-4"
        />
        <div>
          <button
            aria-label="notification"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <div
                class="MuiAvatar-root makeStyles-icon-51 MuiAvatar-colorDefault"
              >
                EN
              </div>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <button
          aria-label="notification"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <span
              class="MuiBadge-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
                />
              </svg>
              <span
                class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary MuiBadge-dot"
              />
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="logout"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M21 3.01H3c-1.1 0-2 .9-2 2V9h2V4.99h18v14.03H3V15H1v4.01c0 1.1.9 1.98 2 1.98h18c1.1 0 2-.88 2-1.98v-14c0-1.11-.9-2-2-2zM11 16l4-4-4-4v3H1v2h10v3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="menu"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          data-testid="menu-icon"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
    <div
      class="makeStyles-container-1"
    >
      <main
        class="makeStyles-main-2"
      >
        <div
          class="makeStyles-coverImage-107"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-lg-4"
            >
              <h2
                class="makeStyles-title-108"
              >
                Praesent sapien massa, convallis a pellentesque nec
              </h2>
              <p
                class="makeStyles-paragraph-109"
              >
                Cras ultricies ligula sed magna dictum porta. Curabitur non nulla sit amet nisl tempus convallis quis ac lectus. Curabitur arcu erat, accumsan id imperdiet et, porttitor at sem. Pellentesque in ipsum id orci porta dapibus. Mauris blandit aliquet elit, eget tincidunt nibh pulvinar a
              </p>
              <div
                class="makeStyles-introButtonContainer-110"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    Explore
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
          </div>
        </div>
        <div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-lg-8"
            >
              <div
                class="makeStyles-featuresContainer-242"
              >
                <h3
                  class="makeStyles-featuresTitle-243"
                >
                  Praesent sapien massa, convallis a
                </h3>
              </div>
              <div
                class="makeStyles-featuresContent-244"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-4"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root MuiPaper-rounded"
                    >
                      <div
                        class="MuiCardMedia-root makeStyles-media-245"
                        style="background-image: url(/static/images/intro1.jpg);"
                        title="Paella dish"
                      />
                      <div
                        class="MuiCardContent-root"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                        >
                          This impressive paella is a perfect party dish and a fun meal to cook together with your guests. Add 1 cup of frozen peas along with the mussels, if you like.
                        </p>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-4"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root MuiPaper-rounded"
                    >
                      <div
                        class="MuiCardMedia-root makeStyles-media-245"
                        style="background-image: url(/static/images/intro2.jpg);"
                        title="Paella dish"
                      />
                      <div
                        class="MuiCardContent-root"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                        >
                          This impressive paella is a perfect party dish and a fun meal to cook together with your guests. Add 1 cup of frozen peas along with the mussels, if you like.
                        </p>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-4"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root MuiPaper-rounded"
                    >
                      <div
                        class="MuiCardMedia-root makeStyles-media-245"
                        style="background-image: url(/static/images/intro3.jpg);"
                        title="Paella dish"
                      />
                      <div
                        class="MuiCardContent-root"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                        >
                          This impressive paella is a perfect party dish and a fun meal to cook together with your guests. Add 1 cup of frozen peas along with the mussels, if you like.
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
          </div>
        </div>
        <div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-lg-8"
            >
              <div
                class="makeStyles-quoteContainer-281"
              >
                <h4
                  class="makeStyles-quote-282"
                >
                  "Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae. Donec velit neque, auctor sit amet aliquam vel, ullamcorper sit amet ligula"
                </h4>
                <div
                  class="makeStyles-signatureContainer-283"
                >
                  <div
                    class="makeStyles-signatureName-284"
                  >
                    Lionel Messi
                  </div>
                  <div
                    class="makeStyles-signatureJob-285"
                  >
                    Greatest footballer of all time
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
          </div>
          <div
            class="makeStyles-coverImage2-286"
          />
        </div>
        <div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-lg-8"
            >
              <div
                class="makeStyles-helpContainer-287"
              >
                <div>
                  <div
                    class="makeStyles-helpIconContainer-288"
                  >
                    <img
                      class="makeStyles-helpIcon-289"
                      src="/static/images/faq.png"
                    />
                  </div>
                  <h4
                    class="makeStyles-helpTitle-290"
                  >
                    Need help?
                  </h4>
                </div>
                <div
                  class="makeStyles-helpInfo-291"
                >
                  Vestibulum ac diam sit amet quam vehicula elementum sed sit amet dui. Curabitur non nulla sit amet nisl tempus convallis quis ac lectus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae
                </div>
                <div
                  class="makeStyles-helpContact-292"
                >
                  support@mindxtech.com
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
            />
          </div>
        </div>
      </main>
      <footer
        class="makeStyles-root-293"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
          />
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-lg-8"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3"
              >
                <h3
                  class="makeStyles-company-294"
                >
                  Company
                </h3>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    About us
                  </a>
                </div>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    Contact
                  </a>
                </div>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    Products
                  </a>
                </div>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    Blogs
                  </a>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3"
              >
                <h3
                  class="makeStyles-company-294"
                >
                  Legal
                </h3>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    FAQ
                  </a>
                </div>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    Premium support
                  </a>
                </div>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    Sponsorships
                  </a>
                </div>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
                  >
                    Custom development
                  </a>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
              >
                <h3
                  class="makeStyles-company-294"
                >
                  Social medias
                </h3>
                <div>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-socialLink-295 MuiTypography-colorInherit"
                  >
                    <img
                      class="makeStyles-image-296"
                      src="/static/images/facebook.png"
                    />
                  </a>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-socialLink-295 MuiTypography-colorInherit"
                  >
                    <img
                      class="makeStyles-image-296"
                      src="/static/images/dribbble.png"
                    />
                  </a>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-socialLink-295 MuiTypography-colorInherit"
                  >
                    <img
                      class="makeStyles-image-296"
                      src="/static/images/google-plus.png"
                    />
                  </a>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-socialLink-295 MuiTypography-colorInherit"
                  >
                    <img
                      class="makeStyles-image-296"
                      src="/static/images/linkedin.png"
                    />
                  </a>
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-socialLink-295 MuiTypography-colorInherit"
                  >
                    <img
                      class="makeStyles-image-296"
                      src="/static/images/pinterest.png"
                    />
                  </a>
                </div>
              </div>
            </div>
            <div
              class="makeStyles-signature-297"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                © 
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                  href="/"
                >
                  Thinh Tran
                </a>
                 
                2019
              </p>
              <span
                class="MuiTypography-root MuiTypography-caption"
              >
                common.footer.createdWithLove
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-lg-2"
          />
        </div>
      </footer>
    </div>
  </div>
  <div
    aria-hidden="true"
    id="simple-menu"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px; visibility: hidden;"
  >
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiMenu-paper MuiPaper-elevation8 MuiPopover-paper MuiPaper-rounded"
      role="document"
      style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiMenu-list MuiList-padding"
        role="menu"
        tabindex="-1"
      >
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          role="menuitem"
          tabindex="0"
        >
          English
          <span
            class="MuiTouchRipple-root"
          />
        </li>
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          role="menuitem"
          tabindex="-1"
        >
          Tiếng Việt
          <span
            class="MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
